### ISSUE TITLE
BelongsTo async should not trigger request when accessing ID

### ISSUE THREAD

**Issue by bakura10** [emberjs/data#2705]
Hi,

I've tried to make a search over the issues, but it seems this issue does not exist yet. Basically, I have a list of invoices that I want to display. The invoices have an async relationship to their payer:

```js
export default DS.Model.extend({
    payer: DS.belongsTo('user', {async: true});
});
```

Then, I display the invoices in a table, and I want to create a link to the payer:

```hbs
{{#link-to 'admin.user' invoice.payer}}{{invoice.payer.id}}{/link-to}}
```

However, this always trigger a new request (and as I show pages of 50 invoices, it does 50 requests). I can't understand the reasoning behind this, as the ID is contained in the invoice, there is no need to trigger requests when what you want is simply creating a link or displaying an id (as it's already loaded in memory).

I'm using Ember-Data beta 14.1, maybe I did something wrong?

**Comment by bmac** [comment]
Hi @bakura10. Unfortunately, there isn't a good way for Ember Data to detect that you want to only access the id of the belongsTo relationship. The problem is Ember Data automatically fetches the async invoice relationship when you do `invoice.get('payer')`. This is useful because users often want to wait on the belongsTo promise to resolve before attempting to manipulate the record. When you do a `{{invoice.payer.id}}` in your template under the hood Ember is doing the equivalent of a `invoice.get('payer').get('id')` so Ember Data never has a change to detect that you are only attempting to access a property that is already loaded.

One work around you could do would be to add a `payerId` attribute to your invoice model. You can then add a custom serializer to duplicate the player property in the json payload so you can access this value as both an attribute and a relationship. I have put together a simple jsbin showing this technique.
http://emberjs.jsbin.com/vaseticufu/4/edit

**Comment by bakura10** [comment]
Hi,

Thank you for the answer, I think setting a simple ID without the explicit belongsTo is indeed the solution. Regarding the issue, I don't know much about Ember-Data internals, but in Doctrine 2 PHP ORM's, we are solving this issue by using a proxy, and the query is done only once the proxy is initialized. And any method fetching the id does not initialize the proxy, so no request. But I understand that constraints are not the same here :).

**Comment by arenoir** [comment]
It seems reasonable to believe if I push normalized payload into a model I should be able to get normalized data out.  Including the foreign keys of belongs to relationships.  

Shouldn't or wouldn't `{{invoice.data.player_id}}` work?

**Comment by bmac** [comment]
`{{invoice.data.player_id}}` would work today. However, I believe the `data` property is not a public api so it could potentially break in the future.

@wecc Has been working on a snapshot API to solve this same issue in the serialization process https://github.com/emberjs/data/pull/2623. Maybe in the future snapshots could be extended to allow access to foreign keys in the template without triggering an ajax request.

**Comment by aaronbhansen** [comment]
I saw the snapshot ticket a few weeks ago, and built a quick implementation to allow access to the ids without triggering a request. We didn't really need access to the the ids in the templates, just in the routes and controllers, but you could build some properties based on it in the controllers to share in the templates. Anyway, here it is, it's worked well for us, and you might be able to use it as a solution until snapshot is released. 
```
DS.Model.reopen
  # returns the belongs to id
  belongsTo: (relation_name) ->
    relationship = @_relationships[relation_name]
    if !(relationship && relationship.relationshipMeta.kind == 'belongsTo')
      throw new Ember.Error("Model '#{Ember.inspect(@)}' has no belongsTo relationship named '#{relation_name}' defined.")
    inverseRecord = Ember.get(relationship, 'inverseRecord')
    Ember.get(inverseRecord, 'id')

  # returns an array of ids
  hasMany: (relation_name) ->
    relationship = @_relationships[relation_name]
    if !(relationship && relationship.relationshipMeta.kind == 'hasMany')
      throw new Ember.Error("Model '#{Ember.inspect(@)}' has no hasMany relationship named '#{relation_name}' defined.")
    ids = []
    members = Ember.get(relationship, 'members');
    members.forEach((member) -> ids.push(Ember.get(member, 'id')))
    ids
```

**Comment by wecc** [comment]
I think this could be closed in favor of #2431

**Comment by igorT** [comment]
This should be supported by this rfc https://github.com/emberjs/rfcs/pull/57

**Comment by justin-hackin** [comment]
Just a workaround, but it took me a while to find one so thought I'd post here:
```
model._internalModel._relationships.initializedRelationships[relationshipName].canonicalState.id
```
was how I got the id of a belongsTo before the associated model was loaded

**Comment by pangratz** [comment]
I think this issue can be closed since an [initial implementation of references](https://github.com/emberjs/data/pull/3303) has been merged into `master`. Having the `ds-references` feature enabled, the id of a `belongsTo` relationship can be accessed without triggering a request to the server via `var familyId = model.belongsTo("family").id()`.

**Comment by fivetanley** [comment]
@pangratz's work on references does indeed close this.

**Comment by erichonkanen** [comment]
it seems like this would be good behavior to have in a mixin or built in, what are thoughts? e.g. right now for every related field on my ED model where I want to use the id but not make a fetch, I have to do this (unless I'm missing a better way!):

```
  /**
   * Caches the creator id to be accessible without doing a fetch
   * for the entire user record for this creator. This prevents
   * excessive server requests when all we want to do is display
   * the users id.
   * (http://emberjs.com/api/data/classes/DS.Model.html#method_belongsTo)
  */
  creator_id: computed('creator', {
    get() {
      const creatorRef = this.belongsTo('creator');
      const rtype = creatorRef.remoteType();
      if (creatorRef && rtype === "link") {
        const creatorLink = creatorRef.link().split('/');
        return creatorLink[creatorLink.length - 1];
      } else if (creatorRef && rtype === "id") {
        return creatorRef.id();
      }
      return;
    }
  }),
```
